tool/release/version/Bump.kt

```diff
@@ -35,34 +26,41 @@ fun bumpVersion(version: String): String {
                     ).toString()
             lastVersionComponent = versionSubComponents.joinToString("-")
         } catch (b: NumberFormatException) {
-            throw RuntimeException("invalid version: ${version}")
+            throw RuntimeException("invalid version: $version")
         }
     }
     versionComponents[versionComponents.lastIndex] = lastVersionComponent
     return versionComponents.joinToString(".")
 }
 
+fun gitCommitAndPush(workspacePath: Path, newVersion: String) {
+    val shell = Shell(logger)
+    shell.execute(listOf("git", "add", "VERSION"), baseDir = workspacePath)
+    shell.execute(listOf("git", "commit", "-m", "\"Bump version number to $newVersion\""), baseDir = workspacePath)
+    val maxRetries = 3
+    var retryCount = 0
+    while (retryCount < maxRetries) {
```

Comment:
We pull first to ensure that the `push` succeeds. If there are merge conflicts, my understanding is that `git` will return a non-zero exit code.

Reply:
Interesting, I suppose it is possible.

What if we simplify the approach to simply always pulling before making the commit:

```
fun gitCommitAndPush(workspacePath: Path, newVersion: String) {
    shell.execute(listOf("git", "pull"), baseDir = workspacePath)
    shell.execute(listOf("git", "add", "VERSION"), baseDir = workspacePath)
    shell.execute(listOf("git", "commit", "-m", "\"Bump version number to $newVersion\""), baseDir = workspacePath)
    shell.execute(listOf("git", "push"), baseDir = workspacePath)
}
```

Reply:
One of the possible failure scenarios is that someone else pushes between the `git pull` and `git push`, causing the push to fail. So I think a retry loop is the most reliable way to handle this admittedly rare case.

Reply:
Hmm I think that's rare enough to justify the need of handling that. Because if we want to be that robust, then for the sake of consistency we will have to be just as robust in every other Bazel target in the repo, mainly for handling network and disk issues, ending up with lots of retry loops.

Therefore for the benefit of code simplicity I'd say we go with the simplified approach. What do you think?

Reply:
We decided to keep the retry logic because someone else pushing is an "expected" error (despite being very rare in this case). We also improved the logic, now checking the exit value of `git push` instead of just allowing it to throw an exception and catching it.